src/sqlancer/StatementReducer.java

```diff
@@ -28,32 +33,59 @@ public void reduce(G state, Reproducer<G> reproducer, G newGlobalState) throws E
         printQueries(knownToReproduceBugStatements);
         System.out.println();
 
-        do {
+        if (knownToReproduceBugStatements.size() <= 1) {
+            return;
+        }
+
+        partitionNum = 2;
+        while (knownToReproduceBugStatements.size() >= 2) {
             observedChange = false;
+
             knownToReproduceBugStatements = tryReduction(state, reproducer, newGlobalState,
-                    knownToReproduceBugStatements, (candidateStatements, i) -> {
-                        candidateStatements.remove((int) i);
+                    knownToReproduceBugStatements, (candidateStatements, s, l) -> {
+                        List<Query<C>> removed = new ArrayList<>();
+                        for (int i = 0; i < candidateStatements.size(); i++) {
+                            if (i >= s + l || i < s) {
+                                removed.add(candidateStatements.get(i));
+                            }
+                        }
+                        candidateStatements.clear();
+                        candidateStatements.addAll(removed);
                         return true;
                     });
-        } while (observedChange);
+            if (!observedChange) {
+                if (partitionNum == knownToReproduceBugStatements.size()) {
+                    break;
+                }
+                // increase the search granularity
+                partitionNum = Math.min(partitionNum * 2, knownToReproduceBugStatements.size());
+            }
+        }
 
         System.out.println("Reduced query:");
         printQueries(knownToReproduceBugStatements);
     }
 
     private List<Query<C>> tryReduction(G state, // NOPMD
             Reproducer<G> reproducer, G newGlobalState, List<Query<C>> knownToReproduceBugStatements,
-            BiFunction<List<Query<C>>, Integer, Boolean> reductionOperation) throws Exception {
+            TriFunction<List<Query<C>>, Integer, Integer, Boolean> reductionOperation) throws Exception {
 
         List<Query<C>> statements = knownToReproduceBugStatements;
-        for (int i = 0; i < statements.size(); i++) {
+
+        int start = 0;
+        int subLength = statements.size() / partitionNum;
+        while (start < statements.size()) {
+            // newStatements = candidate[:start] + candidate[start+subLength:]
+            // in other word, remove [start, start+subLength) from candidates
             try (C con2 = provider.createDatabase(newGlobalState)) {
                 newGlobalState.setConnection(con2);
                 List<Query<C>> candidateStatements = new ArrayList<>(statements);
-                if (!reductionOperation.apply(candidateStatements, i)) {
+                if (!reductionOperation.apply(candidateStatements, start, subLength)) {
```

Comment:
The `reductionOperation.apply(candidateStatements, start, subLength)` always returns true. Even though the old code does the same, need we improve this?

Reply:
Thanks. I think it sould be improved.